### PR TITLE
Refactoring of Server.h: Isolate server management from other logic

### PR DESCRIPTION
**Pull request by TTPO100AJIEX**
Logic to manage protocol servers is implemented in the global Server class, which is responsible for the "main" function. I find it reasonable to abstract this in a separate class responsible only for managing the list of servers to improve readability and code encapsulation.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 2960947e1c2e3e41ae32f123f9600936e8372517 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64132/2960947e1c2e3e41ae32f123f9600936e8372517/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by rschu1ze — COMMENTED** [review]
To be honest, extracting and splitting the code into separate files makes sense if it was used by multiple places but this is not the case here. The only user of "InterServerManager" and "ProtocolServersManager" is Server.cpp. Things are IMHO easier to understand if everything is in one place. I would therefore prefer to not merge this PR (unless there is a good reason which I missed).

**Comment by TTPO100AJIEX** [comment]
> To be honest, extracting and splitting the code into separate files makes sense if it was used by multiple places but this is not the case here. The only user of "InterServerManager" and "ProtocolServersManager" is Server.cpp. Things are IMHO easier to understand if everything is in one place. I would therefore prefer to not merge this PR (unless there is a good reason which I missed).

The idea behind the change was to aid the addition of new protocols - a relatively common improvement that should not require modifications of existing code. Currently, besides implementing the new logic, you would have to change Server.cpp, which is over 2500 lines of code - not the best experience. 
After the change, the code will be encapsulated in ServersManager - a class responsible solely for holding a list of servers, which aids in making the Server responsible for exactly one thing - initialization of all components. With the manager, everything related to running the servers will remain in one (but different) place explicitly separated from everything else happening in the Server. Thus, to add a new protocol, one would need to modify only a much smaller file - ProtocolServersManager.cpp. 
Furthermore, both InterServersManager and ProtocolServersManager have a hard-coded list of servers to run, which I think is a bad architectural decision. Abstracting this logic in separate classes helps later rework this as well.
Finally, there is similar logic in Keeper.cpp: https://github.com/ClickHouse/ClickHouse/blob/master/programs/keeper/Keeper.cpp#L441 . We might find a way to encapsulate all three places behind a common interface, which improves code readability.

**Comment by rschu1ze** [comment]
@TTPO100AJIEX Okay, fine with me then, thanks! The only problem I have at this point is that the PR mixes tons of unrelated formatting/cosmetic changes with actual refactorings (moved code). I am really not against formatting changes, but in order to make this reviewable and also revertable (<-- in case something goes wrong), we need to separate the signal from the noise. Therefore, it would be nice if you could rework the commit history ([commit messages](https://github.com/ClickHouse/ClickHouse/pull/64132/commits) are at the moment trash anyways - excuse me, just stating the obvious), and make two commits: 1. with all formatting changes 2. the actual refactoring - then force-push and remove "[WIP]" from the PR title.

**Comment by TTPO100AJIEX** [comment]
@rschu1ze Could you please have a look at the PR? I have removed the styling changes and squashed the commits. There are some failed tests, but I do not understand how the changes could have affected them (especially the upgrade check). Is there a possibility those tests are inherently broken? Could you please clarify, what exactly those tests are checking?

**Review by rschu1ze — COMMENTED** [review]
Thanks! In general, the refactoring looks like it was done right though I did not step through the code or verify it further.

Please see my (mostly minor) comments, after adressing them we can merge.

File: src/Server/ServersManager/ProtocolServersManager.h
Comment (TTPO100AJIEX) [reply src/Server/ServersManager/ProtocolServersManager.h:24]: I do not think we can do it here. If omitted, the function will not exist on ProtocolServersManager, because it has a virtual override with different parameters.

**Review by rschu1ze — APPROVED** [review]

**Comment by TTPO100AJIEX** [comment]
@rschu1ze The CI has completed. Some tests failed, but they seem unrelated to the changes again (and they are different from what had failed before). Could you please merge the PR as it is?

**Comment by tavplubix** [comment]
@rschu1ze, some CI checks have not finished

**Comment by TTPO100AJIEX** [comment]
> Revert "Refactoring of Server.h: Isolate server management from other logic" #64425

But aren't those checks just broken or flaky? Something fails for all PRs, and the changes do not seem to be related to the failed tests.

**Comment by rschu1ze** [comment]
@TTPO100AJIEX As you probably know, ClickHouse also maintains an internal repository and all "public" PRs are picked in the background into the non-public repo. In this case, there was a merge conflict in the non-public repository (as can be expected by such large refactorings) but I merged your PR yesterday without noticing. By bad, sorry. I'll do a revert of the revert soon and check/resolve these conflicts.

**Comment by javisantana** [comment]
I do think you guys should give a heads up about this kind of things, otherwise it'll be impossible to contribute in a few months (when your fork diverges enough)

**Comment by rschu1ze** [comment]
Sorry for the delay, priorities got a bit disarranged for me this week. I'll definitely check today, promised.

**Pull request merged**
